constraint/solver/gkrgates/registry.go

```diff
@@ -35,78 +37,155 @@ type registerSettings struct {
-	return func(settings *registerSettings) {
+func WithUnverifiedDegree(degree int) RegisterOption {
+	return func(settings *registerSettings) error {
+		if settings.degree != -1 {
+			return fmt.Errorf("gate degree already set to %d", settings.degree)
+		}
 		settings.noDegreeVerification = true
 		settings.degree = degree
+		return nil
 	}
 }
 
 // WithDegree sets the degree of the gate. RegisterGate will return an error if the degree is not correct.
-func WithDegree(degree int) registerOption {
-	return func(settings *registerSettings) {
+func WithDegree(degree int) RegisterOption {
+	return func(settings *registerSettings) error {
+		if settings.degree != -1 {
+			return fmt.Errorf("gate degree already set to %d", settings.degree)
+		}
 		settings.degree = degree
+		return nil
 	}
 }
 
 // WithName can be used to set a human-readable name for the gate.
-func WithName(name gkr.GateName) registerOption {
-	return func(settings *registerSettings) {
+func WithName(name gkr.GateName) RegisterOption {
+	return func(settings *registerSettings) error {
+		if name == "" {
+			return errors.New("gate name must not be empty")
+		}
+		if settings.name != "" {
+			return fmt.Errorf("gate name already set to \"%s\"", settings.name)
+		}
 		settings.name = name
+		return nil
 	}
 }
 
 // WithCurves determines which curves the gate is validated on.
 // The default is to validate on BN254.
 // This works for most gates, unless the leading coefficient is divided by
 // the curve's order, in which case the degree will be computed incorrectly.
-func WithCurves(curves ...ecc.ID) registerOption {
-	return func(settings *registerSettings) {
+func WithCurves(curves ...ecc.ID) RegisterOption {
+	return func(settings *registerSettings) error {
+		if settings.curves != nil {
+			return errors.New("gate curves already set")
+		}
 		settings.curves = curves
+		return nil
 	}
 }
 
 // Register creates a gate object and stores it in the gates registry.
 // - name is a human-readable name for the gate.
 // - f is the polynomial function defining the gate.
 // - nbIn is the number of inputs to the gate.
-func Register(f gkr.GateFunction, nbIn int, options ...registerOption) error {
-	s := registerSettings{degree: -1, solvableVar: -1, name: GetDefaultGateName(f), curves: []ecc.ID{ecc.BN254}}
+//
+// If the gate is already registered, it will return false and no error.
+func Register(f gkr.GateFunction, nbIn int, options ...RegisterOption) error {
+	s := registerSettings{degree: -1, solvableVar: -1}
 	for _, option := range options {
-		option(&s)
+		if err := option(&s); err != nil {
+			return err
+		}
+	}
+	if s.name == "" {
+		s.name = GetDefaultGateName(f)
+	}
+
+	curvesForTesting := s.curves
+	allowedCurves := s.curves
+	if len(curvesForTesting) == 0 {
+		// no restriction on curves, but only test on BN254
+		curvesForTesting = []ecc.ID{ecc.BN254}
+		allowedCurves = gnark.Curves()
+	}
+
+	gatesLock.Lock()
+	defer gatesLock.Unlock()
+
+	if g, ok := gates[s.name]; ok {
```

Comment:
I would keep the duplicate checkins simple -- if there is same duplicate (key, gate) then it is a no-op. If there is same key but different gate then it is not.

It becomes more complicated when the gate functions are anonymous. But imo this means we should make the gate functions simpler and non-anonymous. It also significantly helps reviewing and auditing as we don't have to find the actual defined gates at runtime and instead can analyze them statically.